internal/controller/reinstall/reinstall.go

```diff
@@ -0,0 +1,378 @@
+				reinstallSpec.Generation))
+
+		setReinstallStatusCondition(clusterInstance, *reinstallRequestProcessedCondition)
+
+		if err = patchClusterInstanceReinstallStatus(ctx, r.Client, log, clusterInstance, patch); err != nil {
+			log.Sugar().Errorf("Failed to update reinstall condition '%s'", reinstallRequestProcessedCondition.Type,
+				zap.Error(err))
+			return reconcile.Result{}, err
+		}
+		log.Sugar().Infof("Updated reinstall condition '%s' successfully", reinstallRequestProcessedCondition.Type)
+
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	result = reconcile.Result{}
+	err = nil
+	updateEndtime := false
+
+	defer func() {
+		patch := client.MergeFrom(clusterInstance.DeepCopy())
+		if updateEndtime {
+			clusterInstance.Status.Reinstall.RequestEndTime = metav1.Now()
+			// Add hold annotation for manual intervention (CNF-15719)
+		}
+
+		changed := setReinstallStatusCondition(clusterInstance, *reinstallRequestProcessedCondition)
+		if changed || updateEndtime {
+			updateErr := patchClusterInstanceReinstallStatus(ctx, r.Client, log, clusterInstance, patch)
+			if updateErr != nil {
+				if err == nil {
+					err = updateErr
+				}
+			}
+		}
+	}()
+
+	tasks := []struct {
+		Description string
+		Run         func(context.Context, *zap.Logger, *v1alpha1.ClusterInstance) (reconcile.Result, error)
+	}{
+		{"Validating reinstall request", r.ensureValidReinstallRequest},
+		{"Deleting rendered manifests", r.ensureRenderedManifestsAreDeleted},
+	}
+	for _, task := range tasks {
+		log.Info("Executing task", zap.String("Task", task.Description))
+		result, err = task.Run(ctx, log, clusterInstance)
+		if !result.IsZero() {
+			log.Info("ClusterInstance to be requeued")
+			return
+		}
+		if err != nil {
+			log.Error("Task failed", zap.String("Task", task.Description), zap.Error(err))
+			reinstallRequestProcessedCondition = reinstallRequestProcessedConditionStatus(
+				metav1.ConditionFalse, v1alpha1.Failed,
+				fmt.Sprintf("Encountered error executing task: %s. Error: %v", task.Description, err))
+			updateEndtime = true
+			return
+		}
+		log.Info("Task completed", zap.String("Task", task.Description))
+	}
+
+	log.Info("Finalizing reinstall request")
+	if err = r.finalizeReinstallRequest(ctx, log, clusterInstance); err != nil {
+		return
+	}
+	reinstallRequestProcessedCondition = reinstallRequestProcessedConditionStatus(
+		metav1.ConditionTrue,
+		v1alpha1.Completed,
+		"The reinstall process completed successfully and the cluster is ready to be reprovisioned.")
+
+	return
+}
+
+func isRequestProcessed(clusterInstance *v1alpha1.ClusterInstance) bool {
+
+	reinstallStatus := clusterInstance.Status.Reinstall
+	if reinstallStatus == nil {
+		return false
+	}
+
+	cond := meta.FindStatusCondition(reinstallStatus.Conditions, string(v1alpha1.ReinstallRequestProcessed))
+	if cond == nil {
+		return false
+	}
+
+	if reinstallStatus.ObservedGeneration == clusterInstance.Spec.Reinstall.Generation {
+		return true
+	}
```

Comment:
Should this be before the `FindStatusCondition` call? Seems odd to fail out if the status condition doesn't exist, but then succeed for something unrelated to the status condition if it does exist.